cpp/src/unary/math_ops.cu

```diff
@@ -0,0 +1,343 @@
+
+// rounding functions
+
+struct DeviceCeil {
+    template<typename T>
+    __device__
+    T apply(T data) {
+        return std::ceil(data);
+    }
+};
+
+struct DeviceFloor {
+    template<typename T>
+    __device__
+    T apply(T data) {
+        return std::floor(data);
+    }
+};
+
+struct DeviceAbs {
+    template<typename T>
+    __device__
+    T apply(T data) {
+        return std::abs(data);
+    }
+};
+
+// bitwise op
+
+struct DeviceInvert {
+    // TODO: maybe sfinae overload this for cudf::bool8
+    template<typename T>
+    __device__
+    T apply(T data) {
+        return ~data;
+    }
+};
+
+// logical op
+
+struct DeviceNot {
+    template<typename T>
+    __device__
+    cudf::bool8 apply(T data) {
+        return static_cast<cudf::bool8>( !data );
+    }
+};
+
+
+template<typename T, typename F>
+static void launch(cudf::column_view const& input, cudf::mutable_column_view& output) {
+    cudf::experimental::unary::Launcher<T, T, F>::launch(input, output);
+}
+
+
+template <typename F>
+struct MathOpDispatcher {
+    template <typename T>
+    typename std::enable_if_t<std::is_arithmetic<T>::value, void>
+    operator()(cudf::column_view const& input, cudf::mutable_column_view& output) {
+        launch<T, F>(input, output);
+    }
+
+    template <typename T>
+    typename std::enable_if_t<!std::is_arithmetic<T>::value, void>
+    operator()(cudf::column_view const& input, cudf::mutable_column_view& output) {
+        CUDF_FAIL("Unsupported datatype for operation");
+    }
+};
+
+
+template <typename F>
+struct BitwiseOpDispatcher {
+    template <typename T>
+    typename std::enable_if_t<std::is_integral<T>::value, void>
+    operator()(cudf::column_view const& input, cudf::mutable_column_view& output) {
+        launch<T, F>(input, output);
+    }
+
+    template <typename T>
+    typename std::enable_if_t<!std::is_integral<T>::value, void>
+    operator()(cudf::column_view const& input, cudf::mutable_column_view& output) {
+        CUDF_FAIL("Unsupported datatype for operation");
+    }
+};
+
+
+template <typename F>
+struct LogicalOpDispatcher {
+private:
+    template <typename T>
+    static constexpr bool is_supported() {
+        return std::is_arithmetic<T>::value ||
+               std::is_same<T, cudf::bool8>::value;
+
+        // TODO: try using member detector
+        // std::is_member_function_pointer<decltype(&T::operator!)>::value;
+    }
+
+public:
+    template <typename T>
+    typename std::enable_if_t<is_supported<T>(), void>
+    operator()(cudf::column_view const& input, cudf::mutable_column_view& output) {
+        cudf::experimental::unary::Launcher<T, cudf::bool8, F>::launch(input, output);
+    }
+
+    template <typename T>
+    typename std::enable_if_t<!is_supported<T>(), void>
+    operator()(cudf::column_view const& input, cudf::mutable_column_view& output) {
+        CUDF_FAIL("Unsupported datatype for operation");
+    }
+};
+
+} // namespace detail
+
+std::unique_ptr<cudf::column> 
```

Comment:
Shouldn't the headers for these files be part of this PR also?

Reply:
yes, will be added those